### PR TITLE
confirm deliver result from remote

### PR DESCRIPTION
**Pull request by xiaoch05**

### CODE REVIEW

**Comment by HackFisher** [comment]
Need to merge master due to project structure changes: #782

File: frame/evm/precompile/contracts/encoder/src/lib.rs
Comment (AurevoirXavier) [inline frame/evm/precompile/contracts/encoder/src/lib.rs:21]: For substrate pallet we use `use sp_std::prelude::*` generally.

File: frame/wormhole/issuing/ethereum/src/lib.rs
Comment (HackFisher) [inline frame/wormhole/issuing/ethereum/src/lib.rs:305]: The name is misleading, and the dispatch call only do `deposit_burn_token_event` ?

Comment (xiaoch05) [reply frame/wormhole/issuing/ethereum/src/lib.rs:305]: this call will deposit the burn token event and trigger schedule_mmr_root to request authorities to sign mmr root, and then relay this event to ethereum chain.

File: frame/evm/precompile/contracts/encoder/src/lib.rs
Comment (AurevoirXavier) [inline frame/evm/precompile/contracts/encoder/src/lib.rs:37]: Are these brackets necessary?

File: frame/wormhole/issuing/ethereum/src/lib.rs
Comment (HackFisher) [inline frame/wormhole/issuing/ethereum/src/lib.rs:301]:
```suggestion
		// When user burn their mapped tokens to unlock remote origin token, mapping token factory will use precompile to call this, this call will deposit the burn token event and trigger schedule_mmr_root to request authorities to sign mmr root, and then relay this event to ethereum chain.
```

File: frame/wormhole/issuing/ethereum/src/lib.rs
Comment (HackFisher) [reply frame/wormhole/issuing/ethereum/src/lib.rs:304]:
```suggestion
		pub fn deposit_burn_token_event_from_precompile(
```

File: frame/evm/precompile/contracts/encoder/src/lib.rs
Comment (AurevoirXavier) [inline frame/evm/precompile/contracts/encoder/src/lib.rs:65]: An unnecessary match I think.

File: frame/evm/precompile/contracts/encoder/src/lib.rs
Comment (xiaoch05) [reply frame/evm/precompile/contracts/encoder/src/lib.rs:37]: Yes this can be deleted if we define it the same as in the contract. Here we use this format to describe it as a function. In fact, we should also add the parameters, but for simplicity, we ignore them.

File: frame/wormhole/backing/s2s/src/lib.rs
Comment (AurevoirXavier) [inline frame/wormhole/backing/s2s/src/lib.rs:107]: I think this line doesn't affect anything.

```
#[pallet::metadata(T::AccountId = "AccountId")]
```

File: frame/support/src/lib.rs
Comment (AurevoirXavier) [inline frame/support/src/lib.rs:107]: Docs require.

```
/// [0..4]  bytes ---- reserved
/// [4..8]  bytes ---- laneID
/// [8..16] bytes ---- message nonce
```

File: frame/support/src/lib.rs
Comment (AurevoirXavier) [inline frame/support/src/lib.rs:97]: Could move this doc below.

File: frame/wormhole/issuing/s2s/src/lib.rs
Comment (HackFisher) [inline frame/wormhole/issuing/s2s/src/lib.rs:136]: https://github.com/darwinia-network/darwinia-common/issues/816

Need refactoring, may be in another PR.

**Comment by HackFisher** [comment]
Companion PR for review together: https://github.com/darwinia-network/darwinia-bridge-sol/pull/66

File: frame/wormhole/backing/s2s/src/lib.rs
Comment (AurevoirXavier) [inline frame/wormhole/backing/s2s/src/lib.rs:135]: `Blake2_128Concat` could replace with `Identity`.

`[u8; 16]` could replace with `BridgeMessageId`.

**Comment by AurevoirXavier** [comment]
Empty hooks block can be remove after #805

File: frame/evm/precompile/contracts/encoder/src/lib.rs
Comment (AsceticBear) [inline frame/evm/precompile/contracts/encoder/src/lib.rs:70]: Can we do like the following:

```rs
let digest = |method| { sha3::Keccak256::digest(method) }[..METHOD_DIG_LEN];

_ if method_digest == digest(BURN_AND_REMOTE_UNLOCK_METHOD) => {...}
```

File: frame/evm/precompile/contracts/encoder/src/lib.rs
Comment (hujw77) [reply frame/evm/precompile/contracts/encoder/src/lib.rs:37]: Is the method selector from the contract, if so the parameters should not be omitted

File: frame/evm/precompile/contracts/encoder/src/lib.rs
Comment (xiaoch05) [reply frame/evm/precompile/contracts/encoder/src/lib.rs:70]: I have tried before, but it reports an unknown size error. Any idea?

File: frame/evm/precompile/contracts/encoder/src/lib.rs
Comment (xiaoch05) [reply frame/evm/precompile/contracts/encoder/src/lib.rs:37]: This selector is used to map the caller. And the method is not the real function of the contract.

File: frame/evm/precompile/contracts/encoder/src/lib.rs
Comment (AurevoirXavier) [reply frame/evm/precompile/contracts/encoder/src/lib.rs:70]: I don't know the context.

But basically, there are two ways.

1. reference the return type
2. specify the return type, IMO the `fn digest` could return a generic

Comment (AsceticBear) [reply frame/evm/precompile/contracts/encoder/src/lib.rs:70]: I have tried this way in my local, it works. FYI

```rs
let digest = |param_method, expected_method| -> bool {
	&sha3::Keccak256::digest(param_method)[..METHOD_DIG_LEN] == expected_method
};

_ if digest(method_digest, BURN_AND_REMOTE_UNLOCK_METHOD) => {..}
```

Comment (AurevoirXavier) [reply frame/evm/precompile/contracts/encoder/src/lib.rs:70]: If that then rename to `check_digest`

Comment (AsceticBear) [reply frame/evm/precompile/contracts/encoder/src/lib.rs:70]: Right.

Comment (xiaoch05) [reply frame/evm/precompile/contracts/encoder/src/lib.rs:70]: This way looks good.

**Comment by xiaoch05** [comment]
I test failed at the last time since some lastest modification, please don't merge until I confirmed the issue.
Why I can't change it to draft now?

File: frame/wormhole/issuing/s2s/src/lib.rs
Comment (xiaoch05) [reply frame/wormhole/issuing/s2s/src/lib.rs:138]: Leave this to another PR.

**Comment by xiaoch05** [comment]
local test passed. Please help to review again.

File: frame/evm/precompile/contracts/encoder/src/lib.rs
Comment (AurevoirXavier) [reply frame/evm/precompile/contracts/encoder/src/lib.rs:65]: Simply use if-else here would be more concise?

File: frame/evm/precompile/contracts/encoder/src/lib.rs
Comment (AurevoirXavier) [inline frame/evm/precompile/contracts/encoder/src/lib.rs:91]: Also here

File: frame/wormhole/issuing/ethereum/src/lib.rs
Comment (AurevoirXavier) [inline frame/wormhole/issuing/ethereum/src/lib.rs:316]: Could you document what 8 for?

File: frame/evm/precompile/contracts/encoder/src/lib.rs
Comment (xiaoch05) [reply frame/evm/precompile/contracts/encoder/src/lib.rs:65]: maybe we will have other pallets using this encoder precompile? Not only s2s and ethereum pallets.

File: frame/wormhole/issuing/ethereum/src/lib.rs
Comment (xiaoch05) [reply frame/wormhole/issuing/ethereum/src/lib.rs:316]: It is better to move it to decoder precompile.

**Review by HackFisher — APPROVED** [review]
Agree that this PR require further cleanup, but also good to go if put next changes in the PR to resolve #816 considering that we need a running test bridge for wormhole ui dev.

Need more cleanup for better naming and simplicity.

File: frame/wormhole/backing/s2s/src/lib.rs
Comment (AsceticBear) [inline frame/wormhole/backing/s2s/src/lib.rs:320]: We should not return a const weight number, better to record the read and write times in this function. A example like this:

https://github.com/paritytech/parity-bridges-common/blob/master/modules/token-swap/src/lib.rs#L484

File: frame/support/src/lib.rs
Comment (AsceticBear) [reply frame/support/src/lib.rs:107]: Why not keeping the message-id format consistent with parity-bridges-common, `(lane_id, message nonce)`. What is the purpose of the reserved bytes?

File: frame/evm/precompile/contracts/encoder/src/lib.rs
Comment (AsceticBear) [inline frame/evm/precompile/contracts/encoder/src/lib.rs:134]: Actually, we do not need this `impl`, `match_digest` has no relationship with `DispatchCallEncoder`, right?

File: frame/support/src/lib.rs
Comment (xiaoch05) [reply frame/support/src/lib.rs:107]: Maybe this definition is not so good. But we should not use `(lane_id, message nonce)` directly as in parity-bridges-common. We need a common definition of `message id`. And this `message id` should be generic for sub<>sub， ethereum<>darwina， bsc<>darwinia, etc.
The reserved is not used now. Maybe it will be filled by chainID or some other identities.

File: frame/evm/precompile/contracts/encoder/src/lib.rs
Comment (xiaoch05) [reply frame/evm/precompile/contracts/encoder/src/lib.rs:134]: Yes, a normal function also works.

File: frame/wormhole/backing/s2s/src/lib.rs
Comment (xiaoch05) [reply frame/wormhole/backing/s2s/src/lib.rs:320]: Here, we use a const low number to ensure the caller not failed.
Furthermore, we need a more accurate calculation of this weight. Because the caller restricts this weight. 
Firstly we should not do so many things in the callback.
Secondly, we need to give an accurate weight of this callback.

Comment (AsceticBear) [reply frame/wormhole/backing/s2s/src/lib.rs:320]: Yeap, there are some weight limits for a callback function, we can create an issue about this and fix it later.

**Comment by HackFisher** [comment]
After checking with @xiaoch05 , he would like to address unresolved conversations in next PRs, related issues: #819 #816 .

Merging this for the requirement of wormhole ui dev.

**Comment by AurevoirXavier** [comment]
Would you mind adding some info to describe this PR? @xiaoch05

**Pull request merged**
